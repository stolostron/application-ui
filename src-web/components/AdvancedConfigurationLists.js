/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import ResourceList from './common/ResourceList'
import ResourceTableModule from './common/ResourceTableModuleFromProps'
import { RESOURCE_TYPES } from '../../lib/shared/constants'
import getResourceDefinitions from '../definitions'
import { makeGetVisibleTableItemsSelector } from '../reducers/common'
import QuerySwitcher from './common/QuerySwitcher'

const AdvancedConfigurationLists = (props) => {

  const resourceType = RESOURCE_TYPES.QUERY_APPLICATIONS
  const staticResourceData = getResourceDefinitions(resourceType)
  const getVisibleResources = makeGetVisibleTableItemsSelector(resourceType)

  return (
    <ResourceList
      {...props}
      resourceType={resourceType}
      staticResourceData={staticResourceData}
      getVisibleResources={getVisibleResources}
      modules={[<ResourceTableModule key="deployments" definitionsKey="deploymentKeys" />]}
    >
      <QuerySwitcher
        key='switcher'
        queryParam="resource"
        options={
              [{ id: 'subscription', contents: 'Subscriptions' },
                { id: 'placementrule', contents: 'Placement rules'},
                {id: 'channel', contents: 'Channels'}]}
        defaultOption="subscription"
      />
    </ResourceList>
  )}

export default AdvancedConfigurationLists
