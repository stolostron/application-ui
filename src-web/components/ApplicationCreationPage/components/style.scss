/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2020. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
@import 'node_modules/carbon-components/scss/globals/scss/_colors';
@import 'scss/variables';

.creation-view-controls-labels {
  padding: 0 0.3125rem 1.125rem 0;

  .bx--form-item.bx--checkbox-wrapper:last-of-type {
    margin-bottom: 10px;
  }
}

.clusterSelector-container {
  position: relative;

  // Show Time window config section only when expanded
  .bx--accordion__content {
    display: none;
  }

  .bx--accordion__item--active {
    .bx--accordion__content {
      width: 100%;
      max-width: 645px;
      display: inline-block;
      background-color: white;
      border-left: 5px solid #dee3e6;
      padding: 1rem;
      margin-bottom: 20px;
      box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.1);
    }
  }

  // Cluster label checkbox
  .clusterSelector-checkbox {
    font-weight: 600;
  }

  // Cluster label description
  .labels-descr {
    font-size: 14px;
    margin-bottom: 30px;
  }

  // Cluster label section
  .labels-section {
    display: block;
  }

  .matching-labels-container {
    display: flex;
    margin-bottom: 20px;
  }

  .matching-labels-input {
    max-width: 45%;
    margin-right: 10px;
    overflow: hidden;
  }

  .text-input {
    background-color: $acm_background_color;
  }

  .add-label-btn {
    width: 190px;
    height: 40px;
    padding: 13px;
    text-align: center;
    cursor: pointer;
  }

  .btn-disabled {
    cursor: not-allowed;
  }

  .add-label-btn-icon {
    margin-left: 7px;
    float: left;
  }

  .add-label-btn-text {
    font-weight: 600;
    font-size: 14px;
    color: #3d70b2;
  }

  .remove-label-btn {
    padding: 12px 5px;
    cursor: pointer;
  }
}

.shared-resource-warning {
  display: inline-flex;
  color: $color__blue-90;
  line-height: 2rem;

  svg {
    fill: #e6ae3c;
    width: 18px;
    height: 18px;
  }

  p {
    font-size: 14px;
  }

  p:first-of-type {
    margin-bottom: 0.5rem;
  }

  div:first-of-type {
    margin-right: 5px;
  }
}

.timeWindow-container {
  position: relative;

  // Active/Block mode btns
  .timeWindow-mode-container {
    display: inline-grid;
    margin-bottom: 10px;
  }

  .mode-btn {
    margin: 5px 0;
  }

  // Time window config accordion
  .bx--accordion__title {
    font-weight: 600;
    font-size: 14px;
  }

  // Show Time window config section only when expanded
  .bx--accordion__content {
    display: none;
  }

  .bx--accordion__item--active {
    .bx--accordion__content {
      width: 100%;
      max-width: 645px;
      display: inline-block;
      background-color: white;
      border-left: 5px solid #dee3e6;
      padding: 1rem;
      margin-bottom: 20px;
      box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.1);
    }
  }

  // Time window config - Days of the week section
  .config-days-section {
    margin-bottom: 40px;
  }

  .config-title {
    font-weight: 600;
    font-size: 14px;
    margin-bottom: 10px;
  }

  .config-descr {
    font-size: 14px;
    margin-bottom: 10px;
  }

  .config-days-selector {
    width: 50%;
    display: grid;
    grid-template-columns: auto auto;

    .first-col {
      grid-column-start: 1;
    }

    .second-col {
      grid-column-start: 2;
    }
  }

  // Time window config - Timezone section
  .config-timezone-section {
    margin-bottom: 40px;
  }

  .config-timezone-combo-box {
    width: 290px;
    background-color: $acm_background_color;
  }

  .bx--list-box__label {
    color: #3d70b2;
  }

  // Add bottom border to differenciate user's local timezone from the list
  .bx--list-box__menu-item:first-of-type {
    border-bottom: 1px solid #dfe3e6;
  }

  // Time window config - Start/End time section
  .config-time-section {
    display: block;
  }

  .config-time-container {
    display: flex;
    margin-bottom: 20px;
  }

  .config-start-time {
    float: left;
  }

  .bx--label {
    font-weight: 600;
    font-size: 14px;
    margin-bottom: 10px;
  }

  .bx--time-picker__input-field {
    width: 140px;
    margin-right: 10px;
    background-color: $acm_background_color;
  }

  .bx--select--inline {
    svg.bx--select__arrow {
      position: relative;
    }
  }

  .add-time-btn {
    width: 220px;
    height: 40px;
    padding: 13px;
    text-align: center;
    cursor: pointer;
  }

  .btn-disabled {
    cursor: not-allowed;
  }

  .add-time-btn-icon {
    margin-left: 7px;
    float: left;
  }

  .add-time-btn-text {
    font-weight: 600;
    font-size: 14px;
    color: #3d70b2;
  }

  .remove-time-btn {
    padding: 12px 5px;
    cursor: pointer;
  }
}
