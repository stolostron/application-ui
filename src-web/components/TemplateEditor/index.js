/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
export { default as TemplateEditor } from './TemplateEditor'
