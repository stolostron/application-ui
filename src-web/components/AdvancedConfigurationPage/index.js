/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
// Copyright (c) 2020 Red Hat, Inc.

import React from 'react'
import msgs from '../../../nls/platform.properties'
import { connect } from 'react-redux'
import resources from '../../../lib/shared/resources'
import {
  updateSecondaryHeader,
  clearSuccessFinished
} from '../../actions/common'

import ApplicationDeploymentHighlights from '../ApplicationDeploymentHighlights'
import { withRouter } from 'react-router-dom'
import AdvancedConfigurationLists from '../AdvancedConfigurationLists'

/* eslint-disable react/prop-types */

resources(() => {
  require('./style.scss')
})

const mapDispatchToProps = dispatch => {
  return {
    clearSuccessFinished: () => clearSuccessFinished(dispatch),
    updateSecondaryHeaderFn: (title, tabs, mainButton) =>
      dispatch(
        updateSecondaryHeader(title, tabs, null, null, null, null, mainButton)
      )
  }
}

class AdvancedConfigurationPage extends React.Component {
  componentDidMount() {
    const {
      secondaryHeaderProps,
      updateSecondaryHeaderFn,
      locale
    } = this.props

    updateSecondaryHeaderFn(
      msgs.get(secondaryHeaderProps.title, locale),
      secondaryHeaderProps.tabs,
      secondaryHeaderProps.mainButton
    )
  }

  componentWillUnmount() {
    this.mutateFinished()
  }

  mutateFinished() {
    this.props.clearSuccessFinished()
  }

  render() {
    const { secondaryHeaderProps } = this.props

    return (
      <React.Fragment>
        <ApplicationDeploymentHighlights />
        <AdvancedConfigurationLists
          secondaryHeaderProps={secondaryHeaderProps}
        />
      </React.Fragment>
    )
  }
}

export default withRouter(
  connect(null, mapDispatchToProps)(AdvancedConfigurationPage)
)
