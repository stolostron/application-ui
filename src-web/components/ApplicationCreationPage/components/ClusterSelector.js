/*******************************************************************************
 * Licensed Materials - Property of IBM
 * Copyright (c) 2020 Red Hat, Inc. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import resources from '../../../../lib/shared/resources'
import {
  Accordion,
  AccordionItem,
  Checkbox,
  Icon,
  TextInput
} from 'carbon-components-react'
import { getSourcePath, removeVs } from '../../TemplateEditor/utils/utils'
import Tooltip from '../../TemplateEditor/components/Tooltip'
import msgs from '../../../../nls/platform.properties'
import _ from 'lodash'

resources(() => {
  require('./style.scss')
})

export class ClusterSelector extends React.Component {
  static propTypes = {
    control: PropTypes.object,
    controlId: PropTypes.string,
    handleChange: PropTypes.func,
    locale: PropTypes.string
  };

  constructor(props) {
    super(props)
    this.state = {}
    if (_.isEmpty(this.props.control.active)) {
      if (
        !this.props.control.showData ||
        this.props.control.showData.length === 0
      ) {
        this.props.control.active = {
          mode: true,
          clusterLabelsList: [
            { id: 0, labelName: '', labelValue: '', validValue: false }
          ],
          clusterLabelsListID: 1
        }
      } else {
        //display existing placement rule
        this.props.control.active = {
          mode: true,
          clusterLabelsList: this.props.control.showData,
          clusterLabelsListID: this.props.control.showData.length
        }
      }
    }
    this.props.control.validation = this.validation.bind(this)
  }

  render() {
    if (!this.props.control.active) {
      this.props.control.active = {
        mode: true,
        clusterLabelsList: [
          { id: 0, labelName: '', labelValue: '', validValue: false }
        ],
        clusterLabelsListID: 1
      }
    }
    const { controlId, locale, control } = this.props
    const { name, active, validation = {} } = control
    const modeSelected = active && active.mode === true
    const isReadOnly = _.get(this.props, 'control.showData', []).length > 0

    return (
      <React.Fragment>
        <div className="creation-view-controls-labels">
          <div>
            {name}
            {validation.required ? (
              <div className="creation-view-controls-required">*</div>
            ) : null}
            <Tooltip control={control} locale={locale} />
          </div>

          <div className="clusterSelector-container">
            <Checkbox
              className="clusterSelector-checkbox"
              name="clusterSelector-checkbox"
              checked={modeSelected}
              disabled={isReadOnly}
              id={`clusterSelector-checkbox-${controlId}`}
              labelText={msgs.get(
                'tooltip.creation.app.settings.clusterSelector',
                locale
              )}
              onClick={this.handleChange.bind(this)}
            />

            <Accordion>
              <AccordionItem open={modeSelected} title={''}>
                <div className="clusterSelector-labels-section">
                  <div className="labels-descr">
                    {msgs.get(
                      'creation.app.settings.selectorClusters.config',
                      locale
                    )}
                  </div>

                  <div
                    className="labels-section"
                    id={`clusterSelector-labels-section-${controlId}`}
                  >
                    {this.renderClusterLabels(control, isReadOnly, controlId)}
                    <div
                      className={`add-label-btn ${
                        isReadOnly ? 'btn-disabled' : ''
                      }`}
                      tabIndex="0"
                      role={'button'}
                      onClick={() => this.addLabelToList(control, !isReadOnly)}
                      onKeyPress={this.addLabelKeyPress.bind(this)}
                    >
                      <Icon
                        name="icon--add--glyph"
                        fill="#3d70b2"
                        description=""
                        className="add-label-btn-icon"
                      />
                      <div className="add-label-btn-text">
                        {msgs.get(
                          'creation.app.settings.selectorClusters.prop.add',
                          locale
                        )}
                      </div>
                    </div>
                  </div>
                </div>
              </AccordionItem>
            </Accordion>
          </div>
        </div>
      </React.Fragment>
    )
  }

  validation(exceptions) {
    const { control, locale, controlId } = this.props
    if (_.get(control, 'active.mode', false)) {
      if (Object.keys(control.active.clusterLabelsList).length === 0) {
        //no cluster labels set
        exceptions.push({
          row: 1,
          text: msgs.get('creation.missing.clusterSelector.value', locale),
          type: 'error',
          controlId: `clusterSelector-labels-section-${controlId}`
        })
      }

      control.active.clusterLabelsList.map(item => {
        const { id, labelName, labelValue } = item
        const invalidLabel = !labelName || labelName.length === 0
        const invalidValue = !labelValue || labelValue.length === 0

        // Add exception if no input for labels or values
        if (invalidLabel) {
          exceptions.push({
            row: 1,
            text: msgs.get('creation.missing.clusterSelector.label', locale),
            type: 'error',
            controlId: `labelName-${id}`
          })
        }
        if (invalidValue) {
          exceptions.push({
            row: 1,
            text: msgs.get('creation.missing.clusterSelector.value', locale),
            type: 'error',
            controlId: `labelName-${id}`
          })
        }
      })
    }
  }

  renderClusterLabels = (control, isReadOnly, controlId) => {
    if (!_.get(control, 'active.clusterLabelsList')) {
      return ''
    }
    return (
      control.active &&
      control.active.clusterLabelsList.map(item => {
        const { id, labelName, labelValue, validValue } = item
        const invalidLabel = !labelName || labelName.length === 0
        const exceptionLabel = msgs.get(
          'creation.missing.clusterSelector.label.field'
        )
        const invalidValue = !labelValue || labelValue.length === 0
        const exceptionValue = msgs.get(
          'creation.missing.clusterSelector.value.field'
        )

        if (validValue || id === 0) {
          return (
            <React.Fragment key={id}>
              <div className="matching-labels-container">
                <div className="matching-labels-input">
                  <TextInput
                    id={`labelName-${id}-${controlId}`}
                    invalid={invalidLabel}
                    invalidText={exceptionLabel}
                    name="labelName"
                    className="text-input"
                    labelText={
                      id === 0
                        ? `${msgs.get('clusterSelector.label.field.ui')}*`
                        : ''
                    }
                    value={labelName === '' ? '' : labelName}
                    placeholder={msgs.get(
                      'clusterSelector.label.placeholder.field'
                    )}
                    disabled={isReadOnly}
                    onChange={this.handleChange.bind(this)}
                  />
                </div>
                <div className="matching-labels-input">
                  <TextInput
                    invalid={invalidValue}
                    invalidText={exceptionValue}
                    id={`labelValue-${id}-${controlId}`}
                    name="labelValue"
                    className="text-input"
                    labelText={
                      id === 0
                        ? `${msgs.get('clusterSelector.value.field.ui')}*`
                        : ''
                    }
                    value={labelValue === '' ? '' : labelValue}
                    placeholder={msgs.get(
                      'clusterSelector.value.placeholder.field'
                    )}
                    disabled={isReadOnly}
                    onChange={this.handleChange.bind(this)}
                  />
                </div>

                {id !== 0 ? ( // Option to remove added labels
                  <div
                    id={id}
                    className="remove-label-btn"
                    tabIndex="0"
                    role={'button'}
                    onClick={() => this.removeLabelFromList(control, item)}
                    onKeyPress={this.removeLabelKeyPress.bind(this)}
                  >
                    <Icon
                      name="icon--close--glyph"
                      fill="#3d70b2"
                      className="remove-label-btn-icon"
                    />
                  </div>
                ) : (
                  ''
                )}
              </div>
            </React.Fragment>
          )
        }
        return ''
      })
    )
  };

  addLabelToList = (control, modeSelected) => {
    if (modeSelected) {
      // Create new "label" item
      control.active.clusterLabelsList.push({
        id: control.active.clusterLabelsListID,
        labelName: '',
        labelValue: '',
        validValue: true
      })
      control.active.clusterLabelsListID++

      // Update UI
      this.forceUpdate()
    }
  };

  addLabelKeyPress = e => {
    if (e.type === 'click' || e.key === 'Enter') {
      this.addLabelToList(this.props.control)
    }
  };

  removeLabelFromList = (control, item) => {
    // Removed labels are no longer valid
    control.active.clusterLabelsList[item.id].validValue = false

    // Update UI and yaml editor
    this.forceUpdate()
    this.handleChange({})
  };

  removeLabelKeyPress = e => {
    if (e.type === 'click' || e.key === 'Enter') {
      this.removeLabelFromList(this.props.control, { id: e.target.id })
    }
  };

  handleChange(event) {
    const { control, handleChange } = this.props
    let targetName = ''
    try {
      targetName = event.target.name
    } catch (e) {
      targetName = ''
    }

    if (targetName) {
      const { active } = control
      if (targetName === 'clusterSelector-checkbox') {
        active.mode = event.target.checked ? true : false
      } else {
        const { clusterLabelsList } = active
        const labelID = parseInt(event.target.id.split('-')[1], 0)
        if (targetName === 'labelName') {
          clusterLabelsList[labelID].labelName = event.target.value
        } else if (targetName === 'labelValue') {
          clusterLabelsList[labelID].labelValue = event.target.value
        }
        clusterLabelsList[labelID].validValue = true
      }
    }

    handleChange(control)
  }
}

export default ClusterSelector

export const summarize = (control, controlData, summary) => {
  const { clusterLabelsList } = control.active || {}
  if (
    clusterLabelsList &&
    _.get(control, 'type', '') !== 'hidden' &&
    _.get(control, 'active.mode')
  ) {
    clusterLabelsList.forEach(({ labelName, labelValue }) => {
      if (labelName && labelValue) {
        summary.push(`${labelName}=${labelValue}`)
      }
    })
  }
}

export const reverse = (control, templateObject) => {
  if (!control.active) {
    let matchLabels
    const local = _.get(
      templateObject,
      getSourcePath(
        'PlacementRule[0].spec.clusterSelector.matchLabels.local-cluster'
      )
    )
    if (!local) {
      matchLabels = _.get(
        templateObject,
        getSourcePath('PlacementRule[0].spec.clusterSelector.matchLabels')
      )
      if (!matchLabels) {
        matchLabels = _.get(
          templateObject,
          getSourcePath('PlacementRule[0].spec.clusterLabels.matchLabels')
        )
      }
    }
    if (matchLabels) {
      matchLabels = removeVs(matchLabels)
      if (matchLabels) {
        const clusterLabelsList = Object.entries(matchLabels).map(
          ([labelName, labelValue], id) => {
            return {
              id,
              labelName,
              labelValue,
              validValue: true
            }
          }
        )
        control.active = {
          mode: true,
          clusterLabelsList,
          clusterLabelsListID: clusterLabelsList.length
        }
      }
    } else {
      const clusterLabelsList = [
        { id: 0, labelName: '', labelValue: '', validValue: false }
      ]
      control.active = {
        mode: false,
        clusterLabelsList,
        clusterLabelsListID: 1
      }
    }
  }
}
