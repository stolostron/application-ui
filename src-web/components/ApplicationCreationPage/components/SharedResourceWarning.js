// Copyright (c) 2020 Red Hat, Inc. All Rights Reserved.
'use strict'

import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import { ExclamationTriangleIcon } from '@patternfly/react-icons'
import _ from 'lodash'
import apolloClient from '../../../../lib/client/apollo-client'
import { SEARCH_QUERY_RELATED } from '../../../apollo-client/queries/SearchQueries'
import { RESOURCE_TYPES } from '../../../../lib/shared/constants'
import resources from '../../../../lib/shared/resources'
import msgs from '../../../../nls/platform.properties'
import { withLocale } from '../../../providers/LocaleProvider'

resources(() => {
  require('./style.scss')
})

const getSelfLinks = control => {
  const channel = _.get(control, 'groupControlData', []).find(
    d => d.id === 'channel'
  )
  const selfLinks = _.get(channel, 'content', []).find(
    c => c.id === 'selfLinks'
  )
  return _.get(selfLinks, 'active', {})
}

const getResourceKind = resourceType => {
  let kind
  switch (resourceType.name) {
  case 'HCMSubscription':
    kind = 'Subscription'
    break
  case 'HCMPlacementRule':
    kind = 'PlacementRule'
    break
  }
  return kind
}

const getResourceNameAndNamespace = (resourceType, selfLinks) => {
  let selfLinksKey, selfLinkPlural
  switch (resourceType.name) {
  case 'HCMSubscription':
    selfLinksKey = 'Subscription'
    selfLinkPlural = 'subscriptions'
    break
  case 'HCMApplication':
    selfLinksKey = 'Application'
    selfLinkPlural = 'applications'
    break
  case 'HCMPlacementRule':
    selfLinksKey = 'PlacementRule'
    selfLinkPlural = 'placementrules'
    break
  }
  const selfLink = _.get(selfLinks, selfLinksKey, '')
  const selfLinkExp = new RegExp(
    `namespaces/([^/]*)/${selfLinkPlural}/([^/]*)`
  )
  const matches = selfLink.match(selfLinkExp)
  return matches ? matches.slice(1, 3) : [null, null]
}

const getQuery = (resourceType, name, namespace) => {
  const kind = getResourceKind(resourceType).toLowerCase()
  return {
    filters: [
      { property: 'kind', values: [kind] },
      { property: 'name', values: [name] },
      { property: 'namespace', values: [namespace] }
    ],
    relatedKinds: ['application']
  }
}

const SharedResourceWarning = ({ resourceType, control, locale }) => {
  const [relatedApplications, setRelatedApplications] = useState([])
  const selfLinks = getSelfLinks(control)
  const [applicationNamespace, applicationName] = getResourceNameAndNamespace(
    RESOURCE_TYPES.HCM_APPLICATIONS,
    selfLinks
  )
  const [resourceNamespace, resourceName] = getResourceNameAndNamespace(
    resourceType,
    selfLinks
  )

  useEffect(
    () => {
      if (control.editMode && resourceName && resourceNamespace) {
        const query = getQuery(resourceType, resourceName, resourceNamespace)
        apolloClient
          .search(SEARCH_QUERY_RELATED, {
            input: [query]
          })
          .then(response => {
            const relatedItems =
              _.get(response, 'data.searchResult[0].related') || []
            const relatedApps = _.get(
              relatedItems.find(r => r.kind === 'application'),
              'items',
              []
            )
            if (relatedApps) {
              setRelatedApplications(
                relatedApps
                  .filter(
                    r =>
                      r.name !== applicationName ||
                      r.namespace !== applicationNamespace
                  )
                  .map(r => r.name)
                  .sort()
              )
            }
          })
      }
    },
    [
      control.editMode,
      resourceNamespace,
      resourceName,
      applicationName,
      applicationNamespace
    ]
  )

  return relatedApplications.length ? (
    <div className="shared-resource-warning">
      <div>
        <ExclamationTriangleIcon />
      </div>
      <div>
        <p>
          {msgs.get(
            'editing.app.sharedResourceWarning',
            [getResourceKind(resourceType)],
            locale
          )}
        </p>
        <p>{relatedApplications.join(', ')}</p>
      </div>
    </div>
  ) : null
}

SharedResourceWarning.propTypes = {
  control: PropTypes.object,
  locale: PropTypes.string,
  resourceType: PropTypes.object
}

export default withLocale(SharedResourceWarning)
