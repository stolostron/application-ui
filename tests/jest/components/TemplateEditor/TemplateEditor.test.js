/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/

"use strict";

import React from "react";
import { TemplateEditor } from "../../../../src-web/components/TemplateEditor";
import renderer from "react-test-renderer";
import applicationTemplate from "../../../../src-web/components/ApplicationCreationPage/templates/template.hbs";
import { shallow } from "enzyme";
import _ from "lodash";
import { mount } from "enzyme";

import { controlData, portals } from "../TestingData";

describe("TemplateEditor component", () => {
  it("renders as expected", () => {
    const component = renderer.create(
      <TemplateEditor
        template={applicationTemplate}
        controlData={controlData}
        portals={portals}
        type={"application"}
        title={"creation.app.yaml"}
        locale={"en-US"}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe("on control change function", () => {
  it("renders as expected", () => {
    const wrapper = mount(
      <TemplateEditor
        template={applicationTemplate}
        controlData={controlData}
        portals={portals}
      />
    );
    const evt = {
      target: {
        value: "value-testing"
      },
      selectedItems: ["selectedItems-testing-1", "selectedItems-testing-2"]
    };

    wrapper
      .find("#name")
      .at(0)
      .simulate("change", evt);
    wrapper
      .find("#namespace")
      .at(0)
      .simulate("change", evt);

    wrapper
      .find("#main-")
      .at(0)
      .simulate("click", evt);
  });
});

describe("getResourceJSON function", () => {
  const result = [
    { apiVersion: "v1", kind: "Namespace", metadata: { name: null } },
    {
      apiVersion: "app.k8s.io/v1beta1",
      kind: "Application",
      metadata: { name: null, namespace: null },
      spec: {
        componentKinds: [
          { group: "apps.open-cluster-management.io", kind: "Subscription" }
        ],
        descriptor: {},
        selector: {
          matchExpressions: [{ key: "app", operator: "In", values: [null] }]
        }
      }
    },
    {
      apiVersion: "apps.open-cluster-management.io/v1",
      kind: "PlacementRule",
      metadata: { labels: { app: null }, name: "-placement", namespace: null }
    }
  ];

  it("renders as expected", () => {
    const wrapper = shallow(
      <TemplateEditor
        template={applicationTemplate}
        controlData={controlData}
        portals={portals}
      />
    );
    expect(wrapper.instance().getResourceJSON()).toEqual(result);
  });
});
