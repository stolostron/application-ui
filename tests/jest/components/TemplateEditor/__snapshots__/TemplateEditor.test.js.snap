// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateEditor component renders as expected 1`] = `
<div
  className="creation-view"
>
  <div
    className="creation-view-split-container"
  >
    <div
      className="creation-view-controls-container"
    >
      <div
        className="creation-view-controls"
        onScroll={[Function]}
      >
        <div
          className="content"
        >
          <div
            aria-label="Application settings"
            className="creation-view-controls-title"
            id="main-"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex="0"
            title="Application settings"
          >
            <div
              className="creation-view-controls-note"
            >
              All fields marked with an asterisk (*) are mandatory. Fill out the form or edit the YAML directly.
            </div>
            <div
              className="creation-view-controls-title-normal-container"
            />
          </div>
          <div
            className="creation-view-controls-section"
          >
            <div
              className="creation-view-controls-textbox"
              style={
                Object {
                  "display": "",
                }
              }
            >
              <label
                className="creation-view-controls-textbox-title"
                htmlFor="name"
              >
                Name
                <div
                  className="bx--tooltip__trigger"
                >
                  
                  <div
                    aria-expanded={false}
                    aria-haspopup="true"
                    id=""
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onMouseOut={[Function]}
                    onMouseOver={[Function]}
                    role="button"
                    tabIndex={0}
                  >
                    <svg
                      alt="tooltip"
                      aria-label="tooltip"
                      fillRule="evenodd"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                    >
                      <title>
                        
                      </title>
                      <path
                        d="M8,14.5 C11.5898509,14.5 14.5,11.5898509 14.5,8 C14.5,4.41014913 11.5898509,1.5 8,1.5 C4.41014913,1.5 1.5,4.41014913 1.5,8 C1.5,11.5898509 4.41014913,14.5 8,14.5 Z M8,16 C3.581722,16 0,12.418278 0,8 C0,3.581722 3.581722,0 8,0 C12.418278,0 16,3.581722 16,8 C16,12.418278 12.418278,16 8,16 Z"
                      />
                      <polygon
                        points="9 13 7 13 7 7 9 7"
                      />
                      <circle
                        cx="8"
                        cy="4"
                        r="1"
                      />
                    </svg>
                  </div>
                </div>
              </label>
              <div
                className="bx--form-item"
              >
                <input
                  autoComplete="new-password"
                  className="bx--text-input bx--text__input"
                  disabled={false}
                  id="name"
                  onChange={[Function]}
                  onClick={[Function]}
                  placeholder="Enter {0}"
                  spellCheck={false}
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              className="creation-view-controls-textbox"
              style={
                Object {
                  "display": "",
                }
              }
            >
              <label
                className="creation-view-controls-textbox-title"
                htmlFor="namespace"
              >
                Namespace
                <div
                  className="bx--tooltip__trigger"
                >
                  
                  <div
                    aria-expanded={false}
                    aria-haspopup="true"
                    id=""
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onMouseOut={[Function]}
                    onMouseOver={[Function]}
                    role="button"
                    tabIndex={0}
                  >
                    <svg
                      alt="tooltip"
                      aria-label="tooltip"
                      fillRule="evenodd"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                    >
                      <title>
                        
                      </title>
                      <path
                        d="M8,14.5 C11.5898509,14.5 14.5,11.5898509 14.5,8 C14.5,4.41014913 11.5898509,1.5 8,1.5 C4.41014913,1.5 1.5,4.41014913 1.5,8 C1.5,11.5898509 4.41014913,14.5 8,14.5 Z M8,16 C3.581722,16 0,12.418278 0,8 C0,3.581722 3.581722,0 8,0 C12.418278,0 16,3.581722 16,8 C16,12.418278 12.418278,16 8,16 Z"
                      />
                      <polygon
                        points="9 13 7 13 7 7 9 7"
                      />
                      <circle
                        cx="8"
                        cy="4"
                        r="1"
                      />
                    </svg>
                  </div>
                </div>
              </label>
              <div
                className="bx--form-item"
              >
                <input
                  autoComplete="new-password"
                  className="bx--text-input bx--text__input"
                  disabled={false}
                  id="namespace"
                  onChange={[Function]}
                  onClick={[Function]}
                  placeholder="Enter {0}"
                  spellCheck={false}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="creation-view-controls-container-blurr bottom"
      />
    </div>
  </div>
</div>
`;
